pkg/kv/kvserver/raftlog/entry.go

```diff
@@ -28,6 +28,49 @@ import (
 	"go.etcd.io/etcd/raft/v3/raftpb"
 )
 
+// EncodingVersion determines the RaftCommandEncodingVersion for a given
+// Entry.
+func EncodingVersion(ent raftpb.Entry) (kvserverbase.RaftCommandEncodingVersion, error) {
+	if len(ent.Data) == 0 {
+		// An empty command.
+		return kvserverbase.RaftVersionEmptyEntry, nil
+	}
+
+	switch ent.Type {
+	case raftpb.EntryNormal:
+	case raftpb.EntryConfChange:
+		return kvserverbase.RaftVersionConfChange, nil
+	case raftpb.EntryConfChangeV2:
+		return kvserverbase.RaftVersionConfChangeV2, nil
+	default:
+		return 0, errors.AssertionFailedf("unknown EntryType %d", ent.Type)
+	}
+
+	v := kvserverbase.RaftCommandEncodingVersion(ent.Data[0])
+	switch v {
+	case kvserverbase.RaftVersionStandard:
+	case kvserverbase.RaftVersionSideloaded:
+	default:
+		return 0, errors.AssertionFailedf("unknown command encoding version %d", v)
+	}
+
+	return v, nil
+}
+
+// DecomposeRaftVersionStandard extracts the CmdIDKey and the marshaled
+// kvserverpb.RaftCommand from a slice which is known to have come from a
+// raftpb.Entry of type kvserverbase.RaftVersionStandard.
+func DecomposeRaftVersionStandard(data []byte) (kvserverbase.CmdIDKey, []byte) {
+	return kvserverbase.CmdIDKey(data[1 : 1+kvserverbase.RaftCommandIDLen]), data[1+kvserverbase.RaftCommandIDLen:]
+}
+
+// DecomposeRaftVersionSideloaded extracts the CmdIDKey and the marshaled
+// kvserverpb.RaftCommand from a slice which is known to have come from a
+// raftpb.Entry of type kvserverbase.RaftVersionSideloaded.
+func DecomposeRaftVersionSideloaded(data []byte) (kvserverbase.CmdIDKey, []byte) {
+	return kvserverbase.CmdIDKey(data[1 : 1+kvserverbase.RaftCommandIDLen]), data[1+kvserverbase.RaftCommandIDLen:]
+}
```

Comment:
There is code above which relies on the fact that these funcs are identical. At the same time, there is code that doesn't. For consistency and brevity, make these one function, and be clear which types it is for? This is for "normal entry", right?

Reply:
Good idea, done.